// <auto-generated />
// Built from: hl7.fhir.r4.core version: 4.0.1


namespace fhir_R4
{
  /// <summary>
  /// A sequence that is used as a reference to describe variants that are present in a sequence analyzed.
  /// </summary>
  struct MolecularSequenceReferenceSeq {
    /// <summary>
    /// Unique id for the element within a resource (for internal references). This may be any string value that does not contain spaces.
    /// </summary>
    string _id;
    /// <summary>
    /// There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
    /// </summary>
    size_t _extension_count;
    Extension** _extension;
    /// <summary>
    /// There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
    /// </summary>
    size_t _modifierExtension_count;
    Extension** _modifierExtension;
    /// <summary>
    /// Structural unit composed of a nucleic acid molecule which controls its own replication through the interaction of specific proteins at one or more origins of replication ([SO:0000340](http://www.sequenceontology.org/browser/current_svn/term/SO:0000340)).
    /// </summary>
    CodeableConcept* _chromosome;
    /// <summary>
    /// The Genome Build used for reference, following GRCh build versions e.g. 'GRCh 37'.  Version number must be included if a versioned release of a primary build was used.
    /// </summary>
    string* _genomeBuild;
    /// <summary>
    /// A relative reference to a DNA strand based on gene orientation. The strand that contains the open reading frame of the gene is the "sense" strand, and the opposite complementary strand is the "antisense" strand.
    /// </summary>
    NullableString8* _orientation;
    /// <summary>
    /// Reference identifier of reference sequence submitted to NCBI. It must match the type in the MolecularSequence.type field. For example, the prefix, “NG_” identifies reference sequence for genes, “NM_” for messenger RNA transcripts, and “NP_” for amino acid sequences.
    /// </summary>
    CodeableConcept* _referenceSeqId;
    /// <summary>
    /// A pointer to another MolecularSequence entity as reference sequence.
    /// </summary>
    Reference* _referenceSeqPointer;
    /// <summary>
    /// A string like "ACGT".
    /// </summary>
    string* _referenceSeqString;
    /// <summary>
    /// An absolute reference to a strand. The Watson strand is the strand whose 5'-end is on the short arm of the chromosome, and the Crick strand as the one whose 5'-end is on the long arm.
    /// </summary>
    NullableString8* _strand;
    /// <summary>
    /// Start position of the window on the reference sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
    /// </summary>
    NullableInt32* _windowStart;
    /// <summary>
    /// End position of the window on the reference sequence. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
    /// </summary>
    NullableInt32* _windowEnd;
  };
  /// <summary>
  /// The definition of variant here originates from Sequence ontology ([variant_of](http://www.sequenceontology.org/browser/current_svn/term/variant_of)). This element can represent amino acid or nucleic sequence change(including insertion,deletion,SNP,etc.)  It can represent some complex mutation or segment variation with the assist of CIGAR string.
  /// </summary>
  struct MolecularSequenceVariant {
    /// <summary>
    /// Unique id for the element within a resource (for internal references). This may be any string value that does not contain spaces.
    /// </summary>
    string _id;
    /// <summary>
    /// There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
    /// </summary>
    size_t _extension_count;
    Extension** _extension;
    /// <summary>
    /// There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
    /// </summary>
    size_t _modifierExtension_count;
    Extension** _modifierExtension;
    /// <summary>
    /// Start position of the variant on the  reference sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
    /// </summary>
    NullableInt32* _start;
    /// <summary>
    /// End position of the variant on the reference sequence. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
    /// </summary>
    NullableInt32* _end;
    /// <summary>
    /// An allele is one of a set of coexisting sequence variants of a gene ([SO:0001023](http://www.sequenceontology.org/browser/current_svn/term/SO:0001023)).  Nucleotide(s)/amino acids from start position of sequence to stop position of sequence on the positive (+) strand of the observed  sequence. When the sequence  type is DNA, it should be the sequence on the positive (+) strand. This will lay in the range between variant.start and variant.end.
    /// </summary>
    string* _observedAllele;
    /// <summary>
    /// An allele is one of a set of coexisting sequence variants of a gene ([SO:0001023](http://www.sequenceontology.org/browser/current_svn/term/SO:0001023)). Nucleotide(s)/amino acids from start position of sequence to stop position of sequence on the positive (+) strand of the reference sequence. When the sequence  type is DNA, it should be the sequence on the positive (+) strand. This will lay in the range between variant.start and variant.end.
    /// </summary>
    string* _referenceAllele;
    /// <summary>
    /// Extended CIGAR string for aligning the sequence with reference bases. See detailed documentation [here](http://support.illumina.com/help/SequencingAnalysisWorkflow/Content/Vault/Informatics/Sequencing_Analysis/CASAVA/swSEQ_mCA_ExtendedCIGARFormat.htm).
    /// </summary>
    string* _cigar;
    /// <summary>
    /// A pointer to an Observation containing variant information.
    /// </summary>
    Reference* _variantPointer;
  };
  /// <summary>
  /// Receiver Operator Characteristic (ROC) Curve  to give sensitivity/specificity tradeoff.
  /// </summary>
  struct MolecularSequenceQualityRoc {
    /// <summary>
    /// Unique id for the element within a resource (for internal references). This may be any string value that does not contain spaces.
    /// </summary>
    string _id;
    /// <summary>
    /// There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
    /// </summary>
    size_t _extension_count;
    Extension** _extension;
    /// <summary>
    /// There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
    /// </summary>
    size_t _modifierExtension_count;
    Extension** _modifierExtension;
    /// <summary>
    /// Invidual data point representing the GQ (genotype quality) score threshold.
    /// </summary>
    size_t _score_count;
    NullableInt32** _score;
    /// <summary>
    /// The number of true positives if the GQ score threshold was set to "score" field value.
    /// </summary>
    size_t _numTP_count;
    NullableInt32** _numTP;
    /// <summary>
    /// The number of false positives if the GQ score threshold was set to "score" field value.
    /// </summary>
    size_t _numFP_count;
    NullableInt32** _numFP;
    /// <summary>
    /// The number of false negatives if the GQ score threshold was set to "score" field value.
    /// </summary>
    size_t _numFN_count;
    NullableInt32** _numFN;
    /// <summary>
    /// Calculated precision if the GQ score threshold was set to "score" field value.
    /// </summary>
    size_t _precision_count;
    NullableString8** _precision;
    /// <summary>
    /// Calculated sensitivity if the GQ score threshold was set to "score" field value.
    /// </summary>
    size_t _sensitivity_count;
    NullableString8** _sensitivity;
    /// <summary>
    /// Calculated fScore if the GQ score threshold was set to "score" field value.
    /// </summary>
    size_t _fMeasure_count;
    NullableString8** _fMeasure;
  };
  /// <summary>
  /// An experimental feature attribute that defines the quality of the feature in a quantitative way, such as a phred quality score ([SO:0001686](http://www.sequenceontology.org/browser/current_svn/term/SO:0001686)).
  /// </summary>
  struct MolecularSequenceQuality {
    /// <summary>
    /// Unique id for the element within a resource (for internal references). This may be any string value that does not contain spaces.
    /// </summary>
    string _id;
    /// <summary>
    /// There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
    /// </summary>
    size_t _extension_count;
    Extension** _extension;
    /// <summary>
    /// There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
    /// </summary>
    size_t _modifierExtension_count;
    Extension** _modifierExtension;
    /// <summary>
    /// INDEL / SNP / Undefined variant.
    /// </summary>
    NullableString8* _type;
    /// <summary>
    /// Gold standard sequence used for comparing against.
    /// </summary>
    CodeableConcept* _standardSequence;
    /// <summary>
    /// Start position of the sequence. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
    /// </summary>
    NullableInt32* _start;
    /// <summary>
    /// End position of the sequence. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
    /// </summary>
    NullableInt32* _end;
    /// <summary>
    /// The score of an experimentally derived feature such as a p-value ([SO:0001685](http://www.sequenceontology.org/browser/current_svn/term/SO:0001685)).
    /// </summary>
    Quantity* _score;
    /// <summary>
    /// Which method is used to get sequence quality.
    /// </summary>
    CodeableConcept* _method;
    /// <summary>
    /// True positives, from the perspective of the truth data, i.e. the number of sites in the Truth Call Set for which there are paths through the Query Call Set that are consistent with all of the alleles at this site, and for which there is an accurate genotype call for the event.
    /// </summary>
    NullableString8* _truthTP;
    /// <summary>
    /// True positives, from the perspective of the query data, i.e. the number of sites in the Query Call Set for which there are paths through the Truth Call Set that are consistent with all of the alleles at this site, and for which there is an accurate genotype call for the event.
    /// </summary>
    NullableString8* _queryTP;
    /// <summary>
    /// False negatives, i.e. the number of sites in the Truth Call Set for which there is no path through the Query Call Set that is consistent with all of the alleles at this site, or sites for which there is an inaccurate genotype call for the event. Sites with correct variant but incorrect genotype are counted here.
    /// </summary>
    NullableString8* _truthFN;
    /// <summary>
    /// False positives, i.e. the number of sites in the Query Call Set for which there is no path through the Truth Call Set that is consistent with this site. Sites with correct variant but incorrect genotype are counted here.
    /// </summary>
    NullableString8* _queryFP;
    /// <summary>
    /// The number of false positives where the non-REF alleles in the Truth and Query Call Sets match (i.e. cases where the truth is 1/1 and the query is 0/1 or similar).
    /// </summary>
    NullableString8* _gtFP;
    /// <summary>
    /// QUERY.TP / (QUERY.TP + QUERY.FP).
    /// </summary>
    NullableString8* _precision;
    /// <summary>
    /// TRUTH.TP / (TRUTH.TP + TRUTH.FN).
    /// </summary>
    NullableString8* _recall;
    /// <summary>
    /// Harmonic mean of Recall and Precision, computed as: 2 * precision * recall / (precision + recall).
    /// </summary>
    NullableString8* _fScore;
    /// <summary>
    /// Receiver Operator Characteristic (ROC) Curve  to give sensitivity/specificity tradeoff.
    /// </summary>
    MolecularSequenceQualityRoc* _roc;
  };
  /// <summary>
  /// Configurations of the external repository. The repository shall store target's observedSeq or records related with target's observedSeq.
  /// </summary>
  struct MolecularSequenceRepository {
    /// <summary>
    /// Unique id for the element within a resource (for internal references). This may be any string value that does not contain spaces.
    /// </summary>
    string _id;
    /// <summary>
    /// There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
    /// </summary>
    size_t _extension_count;
    Extension** _extension;
    /// <summary>
    /// There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
    /// </summary>
    size_t _modifierExtension_count;
    Extension** _modifierExtension;
    /// <summary>
    /// Click and see / RESTful API / Need login to see / RESTful API with authentication / Other ways to see resource.
    /// </summary>
    NullableString8* _type;
    /// <summary>
    /// URI of an external repository which contains further details about the genetics data.
    /// </summary>
    NullableString8* _url;
    /// <summary>
    /// URI of an external repository which contains further details about the genetics data.
    /// </summary>
    string* _name;
    /// <summary>
    /// Id of the variant in this external repository. The server will understand how to use this id to call for more info about datasets in external repository.
    /// </summary>
    string* _datasetId;
    /// <summary>
    /// Id of the variantset in this external repository. The server will understand how to use this id to call for more info about variantsets in external repository.
    /// </summary>
    string* _variantsetId;
    /// <summary>
    /// Id of the read in this external repository.
    /// </summary>
    string* _readsetId;
  };
  /// <summary>
  /// Structural variant outer.
  /// </summary>
  struct MolecularSequenceStructureVariantOuter {
    /// <summary>
    /// Unique id for the element within a resource (for internal references). This may be any string value that does not contain spaces.
    /// </summary>
    string _id;
    /// <summary>
    /// There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
    /// </summary>
    size_t _extension_count;
    Extension** _extension;
    /// <summary>
    /// There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
    /// </summary>
    size_t _modifierExtension_count;
    Extension** _modifierExtension;
    /// <summary>
    /// Structural variant outer start. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
    /// </summary>
    NullableInt32* _start;
    /// <summary>
    /// Structural variant outer end. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
    /// </summary>
    NullableInt32* _end;
  };
  /// <summary>
  /// Structural variant inner.
  /// </summary>
  struct MolecularSequenceStructureVariantInner {
    /// <summary>
    /// Unique id for the element within a resource (for internal references). This may be any string value that does not contain spaces.
    /// </summary>
    string _id;
    /// <summary>
    /// There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
    /// </summary>
    size_t _extension_count;
    Extension** _extension;
    /// <summary>
    /// There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
    /// </summary>
    size_t _modifierExtension_count;
    Extension** _modifierExtension;
    /// <summary>
    /// Structural variant inner start. If the coordinate system is either 0-based or 1-based, then start position is inclusive.
    /// </summary>
    NullableInt32* _start;
    /// <summary>
    /// Structural variant inner end. If the coordinate system is 0-based then end is exclusive and does not include the last position. If the coordinate system is 1-base, then end is inclusive and includes the last position.
    /// </summary>
    NullableInt32* _end;
  };
  /// <summary>
  /// Information about chromosome structure variation.
  /// </summary>
  struct MolecularSequenceStructureVariant {
    /// <summary>
    /// Unique id for the element within a resource (for internal references). This may be any string value that does not contain spaces.
    /// </summary>
    string _id;
    /// <summary>
    /// There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
    /// </summary>
    size_t _extension_count;
    Extension** _extension;
    /// <summary>
    /// There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
    /// </summary>
    size_t _modifierExtension_count;
    Extension** _modifierExtension;
    /// <summary>
    /// Information about chromosome structure variation DNA change type.
    /// </summary>
    CodeableConcept* _variantType;
    /// <summary>
    /// Used to indicate if the outer and inner start-end values have the same meaning.
    /// </summary>
    NullableBoolean* _exact;
    /// <summary>
    /// Length of the variant chromosome.
    /// </summary>
    NullableInt32* _length;
    /// <summary>
    /// Structural variant outer.
    /// </summary>
    MolecularSequenceStructureVariantOuter* _outer;
    /// <summary>
    /// Structural variant inner.
    /// </summary>
    MolecularSequenceStructureVariantInner* _inner;
  };
  /// <summary>
  /// Raw data describing a biological sequence.
  /// </summary>
  struct MolecularSequence {
    ResourceType resourceType;
    /// <summary>
    /// The only time that a resource does not have an id is when it is being submitted to the server using a create operation.
    /// </summary>
    string _id;
    /// <summary>
    /// The metadata about the resource. This is content that is maintained by the infrastructure. Changes to the content might not always be associated with version changes to the resource.
    /// </summary>
    Meta* _meta;
    /// <summary>
    /// Asserting this rule set restricts the content to be only understood by a limited set of trading partners. This inherently limits the usefulness of the data in the long term. However, the existing health eco-system is highly fractured, and not yet ready to define, collect, and exchange data in a generally computable sense. Wherever possible, implementers and/or specification writers should avoid using this element. Often, when used, the URL is a reference to an implementation guide that defines these special rules as part of it's narrative along with other profiles, value sets, etc.
    /// </summary>
    NullableString8* _implicitRules;
    /// <summary>
    /// Language is provided to support indexing and accessibility (typically, services such as text to speech use the language tag). The html language tag in the narrative applies  to the narrative. The language tag on the resource may be used to specify the language of other presentations generated from the data in the resource. Not all the content has to be in the base language. The Resource.language should not be assumed to apply to the narrative automatically. If a language is specified, it should it also be specified on the div element in the html (see rules in HTML5 for information about the relationship between xml:lang and the html lang attribute).
    /// </summary>
    NullableString8* _language;
    /// <summary>
    /// Contained resources do not have narrative. Resources that are not contained SHOULD have a narrative. In some cases, a resource may only have text with little or no additional discrete data (as long as all minOccurs=1 elements are satisfied).  This may be necessary for data from legacy systems where information is captured as a "text blob" or where text is additionally entered raw or narrated and encoded information is added later.
    /// </summary>
    Narrative* _text;
    /// <summary>
    /// This should never be done when the content can be identified properly, as once identification is lost, it is extremely difficult (and context dependent) to restore it again. Contained resources may have profiles and tags In their meta elements, but SHALL NOT have security labels.
    /// </summary>
    size_t _contained_count;
    Resource** _contained;
    /// <summary>
    /// There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
    /// </summary>
    size_t _extension_count;
    Extension** _extension;
    /// <summary>
    /// There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
    /// </summary>
    size_t _modifierExtension_count;
    Extension** _modifierExtension;
    /// <summary>
    /// A unique identifier for this particular sequence instance. This is a FHIR-defined id.
    /// </summary>
    size_t _identifier_count;
    Identifier** _identifier;
    /// <summary>
    /// Amino Acid Sequence/ DNA Sequence / RNA Sequence.
    /// </summary>
    NullableString8* _type;
    /// <summary>
    /// Whether the sequence is numbered starting at 0 (0-based numbering or coordinates, inclusive start, exclusive end) or starting at 1 (1-based numbering, inclusive start and inclusive end).
    /// </summary>
    NullableInt32* _coordinateSystem;
    /// <summary>
    /// The patient whose sequencing results are described by this resource.
    /// </summary>
    Reference* _patient;
    /// <summary>
    /// Specimen used for sequencing.
    /// </summary>
    Reference* _specimen;
    /// <summary>
    /// The method for sequencing, for example, chip information.
    /// </summary>
    Reference* _device;
    /// <summary>
    /// The organization or lab that should be responsible for this result.
    /// </summary>
    Reference* _performer;
    /// <summary>
    /// The number of copies of the sequence of interest. (RNASeq).
    /// </summary>
    Quantity* _quantity;
    /// <summary>
    /// A sequence that is used as a reference to describe variants that are present in a sequence analyzed.
    /// </summary>
    MolecularSequenceReferenceSeq* _referenceSeq;
    /// <summary>
    /// The definition of variant here originates from Sequence ontology ([variant_of](http://www.sequenceontology.org/browser/current_svn/term/variant_of)). This element can represent amino acid or nucleic sequence change(including insertion,deletion,SNP,etc.)  It can represent some complex mutation or segment variation with the assist of CIGAR string.
    /// </summary>
    size_t _variant_count;
    MolecularSequenceVariant** _variant;
    /// <summary>
    /// Sequence that was observed. It is the result marked by referenceSeq along with variant records on referenceSeq. This shall start from referenceSeq.windowStart and end by referenceSeq.windowEnd.
    /// </summary>
    string* _observedSeq;
    /// <summary>
    /// An experimental feature attribute that defines the quality of the feature in a quantitative way, such as a phred quality score ([SO:0001686](http://www.sequenceontology.org/browser/current_svn/term/SO:0001686)).
    /// </summary>
    size_t _quality_count;
    MolecularSequenceQuality** _quality;
    /// <summary>
    /// Coverage (read depth or depth) is the average number of reads representing a given nucleotide in the reconstructed sequence.
    /// </summary>
    NullableInt32* _readCoverage;
    /// <summary>
    /// Configurations of the external repository. The repository shall store target's observedSeq or records related with target's observedSeq.
    /// </summary>
    size_t _repository_count;
    MolecularSequenceRepository** _repository;
    /// <summary>
    /// Pointer to next atomic sequence which at most contains one variant.
    /// </summary>
    size_t _pointer_count;
    Reference** _pointer;
    /// <summary>
    /// Information about chromosome structure variation.
    /// </summary>
    size_t _structureVariant_count;
    MolecularSequenceStructureVariant** _structureVariant;
  };
};
